### PR TITLE
MapLoader: Connect rooms with the doors next to them

### PR DESCRIPTION
**Pull request by daniellimws**
* Refactor the map parsing code to consider Wall/Room/Door as a Region. 
* Use a 2D map to store the type of region for each pixel. Then for each pixel, if it is a room, check if any of its neighbouring pixels are a door, and if so, connect the door with the room.
* Room class exposes public methods to open or close doors that are
connected to it.
* Temporarily assign keys H/J to open/close all doors for demo purposes.

### CODE REVIEW

**Comment by daniellimws** [comment]
The player movement might feel weird in some of the rooms. This might be a problem with the collider autogenerated from the csgmesh. We can fix this in a different pr.

**Comment by owenl131** [comment]
LGTM

File: Scripts/Generator/Room.cs [hunk 1]
```diff
@@ -28,6 +32,23 @@ public Room(Vector2[][] polygon, int unitSize, Material material)
 
         AddChild(floorMesh);
         Translation = new Vector3(0, -unitSize, 0);
+
+        floorMesh.UseCollision = true;
+        floorMesh.CollisionLayer = 1;
```
Comment (jeffsieu) [inline Scripts/Generator/Room.cs]: `floorMesh.CollisionLayer = ColLayer.Environment;`

File: Scripts/Generator/Room.cs [hunk 2]
```diff
@@ -28,6 +32,23 @@ public Room(Vector2[][] polygon, int unitSize, Material material)
 
         AddChild(floorMesh);
         Translation = new Vector3(0, -unitSize, 0);
+
+        floorMesh.UseCollision = true;
+        floorMesh.CollisionLayer = 1;
+        floorMesh.CollisionMask = 1;
```
Comment (jeffsieu) [inline Scripts/Generator/Room.cs]: `floorMesh.CollisionMask = ColLayer.Environment;`

**Review by jeffsieu — CHANGES_REQUESTED** [review]

File: Scripts/Generator/Room.cs (hunk 1 above)
Comment (daniellimws) [reply Scripts/Generator/Room.cs]: I alr forgot got this

**Review by jeffsieu — APPROVED** [review]

**Pull request merged**